File: Euler_ODE_TIFR.cpp
/*This is a program to solve initial value problem using Euler's Method

Author:- Souparna Nath

Date:- 19/04/20
*/

#include <stdio.h>
#include <math.h>

double f(double,double);
double y_exact(double);


double f(double t,double y)
{
  return y-t*t + 1;
}


double y_exact(double t)
{
  return (t+1)*(t+1) - 0.5*exp(t);
}


int main()
{
  int a = 0;
  int b = 2;
  double y0 = 0.5;
  double h = 0.01;
  int N = 100*(b-a);

  double y_n[N];
  double err[N];
  double err_bound[N];

  y_n[0] = y0;

  for (int i = 1; i<=N; i++){
    double ynew = y0 + h*f(a+(i-1)*h,y0);
    y_n[i] = ynew;
    y0 = ynew;
  }


  for (int j = 0; j<N; j++){
    err[j] = y_n[j] - y_exact(a+j*h);

    /*Local Error Bound is defined as 1/2*M*dt, where M is the bound on
    y''(t) at t = some t0.
    Now, y''(t) = y' - 2t = y -t^2 -2t +1 = y +2 -(t+1)^2
    (t +1)^2 <= 0, Implies |y''| <= y+2 = M.
    y is known. 
    Hence, y(t)+ 2 = M
    */
    double M = y_exact(a+j*h)+2;
    
    err_bound[j] = 0.5*M*h;
  }


  printf("t\tApproximate Value\tExact Value\tRelative Error\tError Bound\n");
  for (int k = 0; k<N; k++){
    printf("%lf\t%lf\t%lf\t%lf\t%lf\n",k*h,y_n[k],y_exact(k*h),err[k],err_bound[k]);
  }
  return 0;
}

File: EulerODE_TIFR.py
"""This is a program to solve initial value problem using Euler's Method

Author:- Souparna Nath

Date:- 19/04/20
"""
import matplotlib.pyplot as plt
import numpy as np

def f(t,y):
    return y/t - (y/t)**2

a = 1
b = 2
y0 = 1

N = 10*(b-a)

y_n = [y0]

for i in range(1,N+1):
    ynew = y0 + 0.1*f(a+(i-1)*0.1,y0)
    y_n.append(ynew)
    y0 = ynew

y_approx = np.array(y_n)

t = np.arange(1,2.1,0.1)
y_exact = t/(1+np.log(t))

abs_err = np.abs(y_exact - y_approx)
rel_err = abs_err/y_exact

plt.plot(t,y_approx,label = r"Approximate Solution of $y' = \frac{y}{t} - \left(\frac{y}{t}\right)^2$")
plt.plot(t,y_exact,label = "Exact Solution")
plt.legend(shadow = True, framealpha = 1.0)

plt.xlabel("t")
plt.ylabel("y(t)")
plt.title("ODE USING EULER'S METHOD")
plt.grid(True)
plt.show()

plt.plot(t,abs_err,label = "Absolute error")
plt.plot(t,rel_err,label = "Relative Error")
plt.legend(shadow = True, framealpha = 1.0)

plt.xlabel("t")
plt.ylabel("Error(t)")
plt.title("ERROR USING EULER'S METHOD")
plt.grid(True)
plt.show()